addons/product_images/wizard/product_fetch_image_wizard.py

```diff
@@ -0,0 +1,243 @@
+                    'params': {
+                        'type': 'warning',
+                        'title': _("Product images"),
+                        'message': _("Custom search quota exceeded. We will start again tomorrow"),
+                        'next': {'type': 'ir.actions.act_window_close'},
+                    }
+                }
+            elif http_error.response.status_code == requests.codes.forbidden:
+                project_url = re.search(r"(?P<url>https?:\/\/[^\s]+)",
+                            http_error.response.json().get('error').get('message')).group("url")
+                raise UserError(_(
+                    "Custom Search API is not enabled in your Google project. Enable it by "
+                    "visiting %(url)s then retry. If you enabled this API recently, wait a few "
+                    "minutes for the action to propagate to Google's systems and retry.",
+                    url=project_url
+                ))
+            elif http_error.response.status_code == requests.codes.bad_request:
+                raise UserError(_(
+                    "Your API Key or your Search Engine ID is incorrect."
+                ))
+            else:
+                raise
+        except ConnectionError:
+            self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                fields.Datetime.now() + timedelta(hours=1.0)
+            )
+        else:
+            if self.products_to_process[10:]:
+                message = _(
+                    "Products are processed in the background. Images will be updated progressively."
+                )
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger()
+            else:
+                message = _(
+                    "%(matching_image_count)s matching images have been found for "
+                    "%(product_count)s products.",
+                    matching_image_count=len(self.products_to_process.filtered('image_1920')),
+                    product_count=len(self.products_to_process)
+                )
+            return {
+                'type': 'ir.actions.client',
+                'tag': 'display_notification',
+                'params': {
+                    'type': 'success',
+                    'title': _("Product images"),
+                    'message': message,
+                    'next': {'type': 'ir.actions.act_window_close'},
+                }
+            }
+
+    def _get_products_to_process(self, limit=10000):
+        """ Get the list of products that need to be processed
+
+        :param int limit: maximum number of records to return (default: 10000)
+        :return: a list of products that meet the criteria
+        :rtype: list
+        """
+        products_to_process = self.env['product.product'].search([
+            ('image_fetch_pending', '=', 'True'),
+        ], limit=limit).filtered(lambda p: not p.image_1920 and p.barcode)
+        return products_to_process
```

Comment:
When I suggested extracting this method I had in mind that it would be called from 2 different parts of your code. As it turns out that only the cron needs to search for products, it's up to you to keep the method or not.

Reply:
Actually, the method could be adapted a tiny bit to do `products_to_process = self or self.search(...)` and then apply the filter. If you call the method from the action too, it allows filtering out products for which an image/barcode has been unset/set meanwhile.

Reply:
I've adapted the method, its docstring, and the code in consequence. Now, I wonder if it's really necessary to put a limit of 10K in the view since the code already applies this limit anyway (if reached, it will just trigger another job the next day). It could be easier for users that have more than 10K products and want to process all of them at once. @pierre-pde, what's your mind about that?